### ISSUE TITLE
`matrix_locations` for many:many requests

### ISSUE THREAD

**Issue by nilsnolde** [valhalla/valhalla#3843]
Putting in an issue before I forget about it.

Should we allow the  `matrix_location` parameter also for many:many requests with timedistance matrix @dnesbitt61? I think it's as useful as on 1:many, restricting the destinations to a best subset and early abort expansion when they're found.

**Comment by dnesbitt61** [comment]
I use TimeDistanceMatrix to compute potential reroute scenarios - finding time + distance from current location to a set of potential locations ahead on the current route. Using matrix_locations allows finding the closest X out of Y locations. This is also useful for computing an optimal location to "join" a route when not currently on a route.

I can't think of use cases for forming partial results of a many-to-many matrix. These are usually used for TSP (optimizing visiting a set of locations) and I'm not sure partial results would work in these cases. If there are good use cases for forming partial many-to-many matrix results then it may make sense to add support (though I imagine most many-to-many matrix requests use CostMatrix).

**Comment by nilsnolde** [comment]
> If there are good use cases for forming partial many

Hmm, point blank I can’t think of a solid use case😅 Doesn’t mean there is none. And IIRC it‘s really just removing the input validation for that parameter and voila, done (negative diff, yay). 

I guess my point is it doesn’t hurt to allow it, even though true, it’s hard to think of a real use case. Wonder if I had one in mind when I opened this🤔

**Comment by octeep** [comment]
> I use TimeDistanceMatrix to compute potential reroute scenarios - finding time + distance from current location to a set of potential locations ahead on the current route. Using matrix_locations allows finding the closest X out of Y locations. This is also useful for computing an optimal location to "join" a route when not currently on a route.
> 
> I can't think of use cases for forming partial results of a many-to-many matrix. These are usually used for TSP (optimizing visiting a set of locations) and I'm not sure partial results would work in these cases. If there are good use cases for forming partial many-to-many matrix results then it may make sense to add support (though I imagine most many-to-many matrix requests use CostMatrix).

I have a use case where I need to do TSP on a bunch of locations where some pairs of locations might be very far away from each other, above the service limit. If many-to-many can return partial results I can at least still do calculations on locations that are below service limits and just assign arbitrarily huge distance to pairs of locations with errors, and still get a pretty good TSP solution. However, since many-to-many just completely fails if one of the pairs fails, I have to figure out which pair of locations is causing the error and that adds a lot of complexities which could be easily avoided if many-to-many could do partial results in the first place.

**Comment by kevinkreiser** [comment]
i also think it is in general useful to return partial results. not because i have a specific use-case in mind but more of a principle type thing. specifically, it seems to me we are trying to assume a use-case so that we can terminate computation early and return nothing. as principles go it feels more correct to avoid assuming a use-case for endpoints which are more statistical in nature and allow the end user to fiddle with whatever use-case it is they have. so long as we can clearly flag when the results are partial (which in this case we can) i think its more useful to be less prescriptive.

long story short, for APIs that are really just summary or statistical in nature (matrix, locate, map matching) it feels like we should just return what we can and let the end user decide if they can or cant use the result. thats my feeling on it anyway.

**Comment by nilsnolde** [comment]
While I agree with that, I think @octeep is latching onto the wrong issue. It’s about terminating early when having a number of specified results, but sounds like he wants the algo to continue after erroring on one pair. That’s also a long standing but different issue:)

**Comment by kevinkreiser** [comment]
True but everyone else got to weigh in on this side question
